Add `allowEmptyInput`, `cache`, `fix` options to configuration object

mattxwang opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Part of (but does not close) #5142. This PR handles all the `boolean` flags (which should be relatively simple).

> Is there anything in the PR that needs further explanation?

Sorry about the CodeQL alerts - it was an approach that I tried (and chose to abandon) that I forgot to remove 😅 

As well as my note below on `cache`, two quick questions:

1. For the configuration docs, what orders should the options go in? I may have missed a pattern - I was confused since it's not in the same order as the flags.
2. Any tests I should add?

---

Note that I didn't `augmentConfig` for `allowEmptyInput` and `cache`, since we never access them through the `config` object outside of `standalone.js`; including them drops code coverage. That being said, maybe there's a cleaner way to use these options "downstream", and be consistent in augmenting the config?

mattxwang commented:
In #6402, it sounds like `cache` may be more complicated than it first seems. I feel like in this case, I must be missing a test case that addresses the complication in the issue (about per-file config resolution) - is that correct?

mattxwang commented:
Thanks for the quick response @ybiquitous, I think you bring up some really great points!

> ### Should CLI flags win config files?
> 
> For example, assume we have the following config file named `.stylelintrc.json`:
> 
> ```json
> { "cache": true }
> ```
> 
> Then, should we expect the following? Note that `--no-` prefix is supported by [`meow`](https://github.com/sindresorhus/meow).
> 
> ```shell
> # Cache is enabled via config file
> stylelint *.css
> 
> # Cache is enabled via CLI flag
> stylelint *.css --cache
> 
> # Cache is disabled via CLI flag
> stylelint *.css --no-cache
> ```
> 
> My answer to the question above is **yes**. Because people would like to turn on/off caching without editing the config file.

I agree with you here (in other words, the CLI flag takes precedence over the config file)! Just implemented the change + have added tests for each of those three cases.

> ### Should we support per-file configs for properties like `cache`?
> 
> I don't believe such a per-file overriding for global settings makes sense. Lint performance and code maintainability would degrade. Even usability might decrease due to config files everywhere.
> 
> So, I recommend supporting only properties in the root config file. We can just ignore properties in sub-folder configs. Of course, we should clearly state the restriction in the document.

This makes sense to me as well. Do you think there are other config options that fall into this category? The rest of them feel like they could be enabled/disabled per-file.

If the user has set `cache` in a subdirectory - should we error, print a message saying that we're ignoring, proceed quietly, or do something else? If this is an exception rather than a rule, I think informing users may be a good idea.

ybiquitous commented:
> Do you think there are other config options that fall into this category? The rest of them feel like they could be enabled/disabled per-file.

Basically, we should prevent per-file config props from increasing in the future (we have now `overrides`). I'm not sure which config props support per-file configs, though. Anyway, not touching the existing props might be safe not to break users' linting.

> If the user has set `cache` in a subdirectory - should we error, print a message saying that we're ignoring, proceed quietly, or do something else? If this is an exception rather than a rule, I think informing users may be a good idea.

I believe erroring is best, but if hard to implement it, just documenting the limitation might be enough.

ybiquitous commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -223,6 +223,91 @@ describe('standalone cache uses cacheLocation', () => {
+		await removeFile(newFileDest);
+	});
+
+	it('uses cache when no CLI argument is provided', async () => {
```
[suggest] I think relating `standalone()` with CLI may be confusing since `standalone()` is called not only CLI but also Node.js programs. So, can you rewrite the descriptions around CLI or `--cache`?

ybiquitous commented on lib/standalone.js:
```diff
@@ -134,7 +134,7 @@ async function standalone({
 		const returnValue = prepareReturnValue([stylelintResult], maxWarnings, formatterFunction, cwd);
 
-		if (fix && postcssResult && !postcssResult.stylelint.ignored) {
+		if ((fix || config?.fix) && postcssResult && !postcssResult.stylelint.ignored) {
```
[suggest] I think the [`??`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) operator can clear the code, like this:

```suggestion
		const autofix = fix ?? config?.fix ?? false;

		if (autofix && postcssResult && !postcssResult.stylelint.ignored) {
```

Note: We can use this operator since Node.js 14 supports it.

ybiquitous commented on lib/standalone.js:
```diff
@@ -164,6 +164,9 @@ async function standalone({
 	}
 
+	// do not cache if config is explicitly overrided by option
+	const useCache = cache || (config?.cache && cache !== false);
```
[suggest] I have the same suggestion as above:

```suggestion
	const useCache = cache ?? config?.cache ?? false;
```

ybiquitous commented on lib/standalone.js:
```diff
@@ -246,7 +249,7 @@ async function standalone({
 
 		stylelintResults = await Promise.all(getStylelintResults);
-	} else if (allowEmptyInput) {
+	} else if (allowEmptyInput || config?.allowEmptyInput) {
```
[suggest] I have the same suggestion as above:

```suggestion
	} else if (allowEmptyInput ?? config?.allowEmptyInput) {
```

ybiquitous commented on lib/standalone.js:
```diff
@@ -30,7 +30,7 @@ const writeFileAtomic = require('write-file-atomic');
  */
 async function standalone({
 	allowEmptyInput = false,
```
Oh, tests fail. This change also is needed:

```suggestion
	allowEmptyInput,
```

mattxwang commented:
Apologies for the delay, got held up with some other work. I think I've addressed most of the code review comments!

To clarify, when you say

> Anyway, not touching the existing props might be safe not to break users' linting.

Do you mean the minimal-invasive change (ex, right now - we follow the "default" behaviour", and I've added a little note in the docs)? Or, do you mean that I should explicitly add a subdirectory check?

ybiquitous commented:
I think we should add some actions for the new config props in this PR, but I don't think they are needed for the existing props.

ybiquitous commented on docs/user-guide/configure.md:
````diff
@@ -446,3 +446,47 @@ If the globs are absolute paths, they are used as is. If they are relative, they
+}
+```
+
+Note: this config option should not be overridden on a per-file basis.
````
[ask] Can we add this note also to `allowEmptyInput` and `fix`?

ybiquitous reviewed (approved):
Thanks, LGTM 👍🏼

ybiquitous commented:
Let's merge this and include it in the next release. 👍🏼 

I've commented on the reason: https://github.com/stylelint/stylelint/issues/6795#issuecomment-1517018386
